### ISSUE TITLE
Terrain textures not loaded with RTSS Terrain

### ISSUE THREAD

**Issue by tritonas00** [OGRECave/ogre#2701]
#### System Information
- Ogre Version: 13.6.0
- Operating System / Platform: Arch linux
- RenderSystem: OpenGL 3+

#### Detailled description
As the title says, terrain textures are not loaded. It started when we migrated to `TerrainMaterialGeneratorA` and RTSS Terrain.

![screenshot_2022-12-22_16-44-22_1](https://user-images.githubusercontent.com/2660424/209158981-618fe611-2b4e-4238-9be8-a17f7ce6b657.png)

It should look like:
![203589552-806f2776-eb65-4c95-a0c6-94d501f3ebe8](https://user-images.githubusercontent.com/2660424/209157989-f66da168-5dc5-4d7a-8945-deeba9d3bc7b.png)

#### RoR.log (includes Ogre log)
[RoR.log](https://github.com/OGRECave/ogre/files/10287665/RoR.log)

#### RenderDoc v1.24 capture
https://www.dropbox.com/s/ieahzjjnq1shm33/cap.rdc?dl=1

**Comment by paroj** [comment]
unfortunately, the vertex shader uniforms are still zero for me, so I get no proper rendering.
However, from looking at (PipelineState > FS), I can tell that all textures are bound and read by the shader.

What catches my eye is that
1. you bind "blank_NRM.dds" to disable layer normal mapping. This massively degrades performance and halves the amount of terrain layers you can use. With the RTSS terrain it is sufficient to just not do this: https://github.com/OGRECave/ogre/blob/9b26b194a330ebeb53c67c87ece4790862fe1999/Samples/EndlessWorld/include/EndlessWorld.h#L282
2. The terrain is not using PSSM. You render 3 shadowmaps, but only the first one is ever used.

**Comment by tritonas00** [comment]
About 2: https://github.com/tritonas00/rigs-of-rods/blob/7711d09bb8e06c31d6c102b3d8d4503b87d3fefb/source/main/gfx/ShadowManager.cpp#L69 this does not apply to the Terrain also?

**Comment by paroj** [comment]
no, terrain needs https://github.com/OGRECave/ogre/blob/0df846d58dd7656cde89cb109eca800cad773e32/Samples/Simple/include/Terrain.h#L627

**Comment by paroj** [comment]
noticed that the VS does not write any texcoords

**Comment by tritonas00** [comment]
Perfect!

![screenshot_2022-12-22_19-44-57_1](https://user-images.githubusercontent.com/2660424/209194992-3cb9e3cc-41ec-4f25-9554-8ebc61a1426b.png)

**Comment by tritonas00** [comment]
Just to let you know, that fixed also the noise artifacts in the terrain i was getting with OpenGL.

**Comment by only-a-ptr** [comment]
> @paroj What catches my eye is that ... you bind "blank_NRM.dds" to disable layer normal mapping. ... With the RTSS terrain it is sufficient to just not do this: `defaultimp.layerList[0].textureNames.push_back("... normheight.dds")`.

Thanks for the hint, I thought normal mapping was pretty cheap novadays. Fact is, that binding comes from a configuration file of the particular terrain asset used here. Our [terrain layer configuration format](https://docs.rigsofrods.org/terrain-creation/terrn2-subsystem/#layer-definitions) requires a normalheight map for every layer, and it seems the terrain authors didn't have any so they used a placeholder. The way I understand your post, we do need a placeholder but only for the layer 0, is that correct?

**Comment by paroj** [comment]
no, do not just use any placeholders

massive degradation might have been a bit too strong for desktop GPUs. It does halve the FPS on my smartphone though. The bad part are the texture fetches and not so much the actual shader code.
However it adds up, given that the terrain is visible on most pixels on the screen and this is per pixels..
